pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,237 @@
+	return &cachingUnstructuredExtractor{
+		dc:    dc,
+		cache: cache,
+		ctx:   ctx,
+	}, nil
+}
+
+// Extract extracts the applied configuration owned by fieldManager from an unstructured object.
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+func discoveryPaths(ctx context.Context, rc rest.Interface) (map[string]OpenAPIGroupVersion, error) {
+	data, err := rc.Get().
+		AbsPath("/openapi/v3").
+		Do(ctx).
+		Raw()
+
+	if err != nil {
+		return nil, err
+	}
+
+	discoMap := &handler3.OpenAPIV3Discovery{}
+	err = json.Unmarshal(data, discoMap)
+	if err != nil {
+		return nil, err
+	}
+
+	oapiPathsToGV := map[string]OpenAPIGroupVersion{}
+	for path, oapiGV := range discoMap.Paths {
+		parse, err := url.Parse(oapiGV.ServerRelativeURL)
+		if err != nil {
+			return nil, err
+		}
+		useClientPrefix := strings.HasPrefix(oapiGV.ServerRelativeURL, "/openapi/v3")
+		etag := parse.Query().Get("hash")
+		oapiPathsToGV[path] = newCustomOAPIGroupVersion(rc, oapiGV, useClientPrefix, etag)
+	}
+	return oapiPathsToGV, nil
+}
+
+// getParserForGV fetches the *GVKParser for the given GroupVersion.
+func (e *cachingUnstructuredExtractor) getParserForGV(ctx context.Context, gv schema.GroupVersion) (*GvkParser, error) { //nolint:gocyclo // for atomic cache operations
+	// parse discovery information
+	oapiPathsToGV, err := discoveryPaths(ctx, e.dc.RESTClient())
+	if err != nil {
+		return nil, err
+	}
+
+	e.cache.mu.Lock()
+	defer e.cache.mu.Unlock()
+	// invalidate stale entries in cache with the fresh discovery data
+	for gvCached, cacheEntry := range e.cache.store {
+		path := gvRelativeAPIPath(gvCached)
+		if discoGV, ok := oapiPathsToGV[path]; !ok || discoGV.ETag() != cacheEntry.etag {
+			delete(e.cache.store, gvCached)
+		}
+	}
+
+	gvPath := gvRelativeAPIPath(gv)
+	oapiGV, ok := oapiPathsToGV[gvPath]
+	if !ok {
+		return nil, fmt.Errorf("cannot find GroupVersion %q in discovery", gvPath)
+	}
+
+	// check the cache after invalidating stale data
+	parserTuple, ok := e.cache.store[gv]
+	// cache hit
+	if ok && parserTuple.etag == oapiGV.ETag() && oapiGV.ETag() != "" {
+		return parserTuple.parser, nil
+	}
+	// generate new parser on cache miss or etag mismatch
+	// defensively cover the case where discovery does not return any ETag
+	// for GV, which normally should not happen
+	freshParser, err := newParserFromOpenAPIGroupVersion(ctx, oapiGV)
```

Comment:
is this expensive? I.e. is it doing any client requests? If yes, we are blocking the cache here.

In other words: why is the mutex a RWMutex if we don't use that?

Reply:
Yes, it is doing a client request to `/openapi/v3/group/version` to retrieve a single GV schema.

Again, it was implemented with a `RLock()` then `Lock()` when cache miss. Then, intentionally switched to single Lock per request at review, to keep all cache related operations in single critical section accepting the performance penalty, assuming cache misses are relatively rare after first call.
I'll switch to regular mutex to keep it consistent.

Reply:
Not good. If there are external requests, this can block. Seconds, if you are unlucky and there is no timeout: minutes.

Reply:
done, https://github.com/crossplane-contrib/provider-kubernetes/pull/246/commits/e8b85762a2218c877855bfba9c28cd5d4bb401a6 uses RLock() to check first, then uses `singleflight` to group schema fetch requests, and ensures only one in-flight request per GV. Concurrent requests are therefore deduplicated, also this happens outside of lock.